stdlib/strings.ncl

```diff
@@ -51,43 +121,305 @@
+
+    replaceRegex: Str -> Str -> Str -> Str
+    | doc m#"
+      Given a string s, regular expression p, and and substring r, replaces every match of p in s with r.
+
+      For example:
+      ```nickel
+        replaceRegex "Hello!" "l+." "j" =>
+          "Hej!"
+        replaceRegex "This 37 is a number." "\\d+" "\"a\" is not" =>
+          "This "a" is not a number."
+      ```
+      "#m
+    = fun s pattern replace =>
        %strReplaceRegex% s pattern replace,
 
-    isMatch : Str -> Str -> Bool = fun s regex => %strIsMatch% s regex,
-    match : Str -> Str -> {match: Str, index: Num, groups: List Str} =
-     fun s regex => %strMatch% s regex,
+    isMatch : Str -> Str -> Bool
+    | doc m#"
+      Checks if the string matches the regular expression.
+
+      For example:
+      ```nickel
+        isMatch "123" "^\\d+$" =>
+          true
+        isMatch "123" "\\d{4}" =>
+          false
+      ```
+      "#m
+    = fun s regex => %strIsMatch% s regex,
+
+    match : Str -> Str -> {match: Str, index: Num, groups: List Str}
+    | doc m#"
+      Matches a string given a regular expression. Results in the part of the string that matched, the index of the
+      first character that was part of the match in the original string, and a lists of all capture groups if any.
+
+      For example:
+      ```nickel
+        match "5 apples, 6 pears and 0 grapes" "^(\\d).*(\\d).*(\\d).*$" =>
+          { match = "5 apples, 6 pears and 0 grapes", index = 0, groups = [ "5", "6", "0" ] }
+        match "01234" "3" =>
+          { match = "3", index = 3, groups = [ ] }
+      ```
+      "#m
+    = fun s regex => %strMatch% s regex,
+
+    length : Str -> Num
+    | doc m#"
+      Results in the length of the given string.
+
+      For example:
+      ```nickel
+        length "" =>
+          0
+        length "hi" =>
+          2
+      ```
+      "#m
+    = fun s => %strLength% s,
+
+    substring: Str -> Num -> Num -> Str
+    | doc m#"
+      Takes a slice from the string. Errors if either index is out of range.
+
+      For example:
+      ```nickel
+        substring "abcdef" 3 5 =>
+          "de"
+        substring "abcdef" 3 10 =>
+          error
+        substring "abcdef" (-3) 4 =>
+          error
+      ```
+      "#m
+    = fun s start end => %strSubstr% s start end,
 
-    length : Str -> Num = fun s => %strLength% s,
+    fromNum | Num -> Str
+    | doc m#"
+      Converts a number to its string representation.
 
-    substring: Str -> Num -> Num -> Str = fun s start end =>
-      %strSubstr% s start end,
+      fromNum 42 =>
+        "42"
+      "#m
+    = fun n => %toStr% n,
 
-    fromNum | Num -> Str = fun n => %toStr% n,
     // fromEnum | < | Dyn> -> Str = fun tag => %toStr% tag,
-    fromEnum | Dyn -> Str = fun tag => %toStr% tag,
-    fromBool | Bool -> Str = fun b => %toStr% b,
+    fromEnum | Dyn -> Str
+    | doc m#"
+      Converts any basic value to its string representation.
```

Comment:
It's technically true, but it's an implementation detail: this function is for enums only. We should probably make it an error to use it with anything else than enum by using a more specific contract for the input. For the time being, let's just lie in the documentation :stuck_out_tongue: